[WIP] Template Support

kevingimbel opened the pull request:
This pull requests adds a `template` functionality to fakedata. The specs and some discussion can be seen in #45. 

The new template functionality supports CLI pipes ([demo asciinema here](https://asciinema.org/a/52lf3qkj6f85aatkccvp7hefn)) and reading in a template file ([demo asciinema here](https://asciinema.org/a/2zkz2tab6hhkkprki6rv92i53)).

As of now this pull request is marked as _Work in Progress_ (WIP) because work is not yet done.

kevingimbel commented on pkg/fakedata/template.go:
```diff
@@ -0,0 +1,171 @@
+
+// ParseTemplate takes a path to a template file as argument. It parses the template file and executes it on
+// os.Stdout.
+func ParseTemplate(path string, limit int) (err error) {
```
@lucapette do you have an idea how we could test this function? It needs a file on disk. Should we provide a file and if so, where would we place it? Together with the golden files for the integration tests?

kevingimbel commented:
@lucapette could you do a review and answer my question regarding testing with template files? Beside the test I'm happy with the current implementation. I think it works quite well and can be merged into master.

lucapette commented:
@kevingimbel I have very limited access to the internet for a few days more. I'm sorry this is waiting for me. I'll do my best to review it as soon as I can!

kevingimbel commented:
@lucapette good to know! I'll start writing the docs if I get the time to do so and maybe change a few bits in the implementation here and there.

kevingimbel commented:
I switched from using `html/template` to `text/template`. The `html/template` package has build-in escaping etc. for working with HTML files, but that is actually a disadvantage for us.

When you have `<` in a custom template it gets escaped and can produce issues, e.g.:

```sh
echo "#{{ Int 1 100}} {{ Name }} <{{ Email }}>" | fakedata
```

will produce the following output: 
```
#45, Gaylord Phillips &lt;giancarlon@example.rehab>
```

You can see that `<` has turned to `&lt;` because it was escaped.

---

As far as I can see there is not much difference between `html/template` and `text/template` beside the added security/escaping when working wit _real_ HTML templates. In my opinion it's safer to use `text/template` for our use case.

lucapette commented:
@kevingimbel something is wrong with commenting (as I did answer the question about testing with templates). 

At first sight, there's only the minor things I've commented (probably it's a good idea to merge master here as there's a conflict).

lucapette commented:
@kevingimbel OK I've just realised there's something else I didnt' notice. The parse methods are printing too (which makes the naming somehow a bit wrong), the problem with that is that now we have the same loop in multiple places. I don't really mind the duplication (as it's really little code) but  I think the code would benefit from reorganizing things in a way that the parse methods and the GenerateRow return  "something to loop *limitFlag times on and print to stdout". If that's not clear (I'm short on time/connection sadly...), we can postpone that and I can take care of it after merge. I'm almost "thinking out loud" right now :)

kevingimbel commented:
@lucapette I think I understand what you mean with the double loop/print and how to better handle it - I'll give this some thought.

As for the merge with master: Yes! I agree it's about time to update this branch with the latest changes from master. I'll do this as soon as I find the time.

kevingimbel commented:
I merged master into this feature branch and resolved the conflicts in `main.go`. 

Next I will re-think the function naming / implementation to make it more readable since `ParseTemplate` also executes the template.

kevingimbel commented:
Of course I forgot to adjust the golden file(s) for integration tests. 🙄

Test should now succeed.

kevingimbel commented:
The template feature is now documented inside the README.md file. Could you proof read what I wrote? Anything missing? @lucapette @jorinvo

kevingimbel commented:
I re-wrote the `ParseTemplate` and `ParseTemplateFromPipe` to only parse the tempalte and return `(template *template.Template, error)`. The previously private method `executeTemplate` is now exported and takes a (parsed) template as well as a limit and then loops and writes to `os.Stdout`. (See https://github.com/lucapette/fakedata/pull/49/commits/1a00f08bcb26927bdffb46583e9eea460d1d5403)

lucapette commented:
@kevingimbel the README additions look good to me at first sight! I think it's a great starting point for the rewriting of the README I plan to do as soon as all the work in progress we have right now lands into master.

The build is still red (errcheck is always checking on us! I like that :)). Apart from the red build, there are just a couple of minor things that I pointed out in comments I'd like to address before we merge (I can take care of it of course if you don't want to/don't have time). 

This is going to be a great feature! 😍

kevingimbel commented:
> Apart from the red build, there are just a couple of minor things that I pointed out in comments

In your previous comments or did you add new comments? When I look at the commits I don't see any new comments.

lucapette commented on main.go:
```diff
@@ -49,13 +51,20 @@ func generatorsHelp() string {
 }
 
+func isPipe() bool {
+	stat, _ := os.Stdin.Stat()
```
why wouldn't we check the error here? (Honest question, I don't know the answer)

kevingimbel replied:
I didn't check during testing if the method works and then forgot to handle it. I'll add error checking and exit if there is an error with reading pipes (makes sense).

lucapette commented on main.go:
```diff
@@ -49,13 +51,20 @@ func generatorsHelp() string {
 		formatFlag     = flag.StringP("format", "f", "", "generators rows in f format. Available formats: csv|tab|sql")
 		versionFlag    = flag.BoolP("version", "v", false, "shows version information")
 		tableFlag      = flag.StringP("table", "t", "TABLE", "table name of the sql format")
+		templateFlag   = flag.StringP("template", "", "", "Use template as input")
```
I think I'd add `-T` as a shortcut

kevingimbel replied:
Will do! I tried `-t` but that's the one for `--table`. `-T` should work.

lucapette commented on main.go:
```diff
@@ -73,13 +82,46 @@ func main() {
+		tmp = t
+	}
+	// Execute a template if there is one
+	if tmp != nil {
```
maybe let's call `tmp` `tmpl`? `tmp` always reminds me of temporary and I think (my guess though) it's a common association

kevingimbel replied:
I agree that `tmp` sounds more like temporary. However, `tmpl` throws a misspell warning with the Go tools (and with my IDE). Maybe we need to go with `t`? I'm currently not sure if misspells break ci or not.

lucapette commented on pkg/fakedata/template.go:
```diff
@@ -0,0 +1,171 @@
+		if i%2 != 0 {
+			return true
+		}
+		return false
```
`return i%2 !=0` should do it right? (I didn't check though)

kevingimbel replied:
Yah that should work. No idea why I went with `if` here. 😂

lucapette commented:
@kevingimbel I think I still don't really know how to use the review feature here. I have to study it a bit. So what I did is adding again all the comments I left (plus some other minor things). It's all small things but I think worth tackling.

I remember we discussed adding integration tests here (the feature calls for that kind of testing in my view) but I'm not sure what's the status of that. Do you still want to take care of that? If not we can still merge this and then I take care of it in a separate pull request (I'm fine with both options so no strings attached).

kevingimbel commented:
@lucapette 

> I think I still don't really know how to use the review feature here. I have to study it a bit. So what I did is adding again all the comments I left (plus some other minor things). It's all small things but I think worth tackling.

It's worth the time! GitHub's Review tool is pretty good in my opinion. Now I see all your comments and I've answered them.

> I remember we discussed adding integration tests here (the feature calls for that kind of testing in my view) but I'm not sure what's the status of that. Do you still want to take care of that?

I agree it screams for integration tests but I have no idea how I would add them and how I would test generation with templates, etc. 

> If not we can still merge this and then I take care of it in a separate pull request (I'm fine with both options so no strings attached).

However you want to handle it. I need to read up on integration tests before I can implement them because I've never done it on my own - I assume I'd add tests to the cli tests and some golden files, but I have seriously no idea how I cans/should tests the templates.

kevingimbel commented:
@lucapette I fixed some things in https://github.com/lucapette/fakedata/pull/49/commits/2992b976b6c3d3ad63be7ecb959444c18d2509f9

- use `-T` as template shortcode
- renamed `tmp` variable to `tmpl`
- remove `if/else` from `Odd` and `Even` funcs
- handle error in `isPipe`

Additionally: 
- Handle file read error in `File` template func
- Add test for non-existing file

---

From my point of view only the Integration tests needs to be done, otherwise the template feature is ready. 🎉

lucapette commented:
@kevingimbel thank you for all the changes! It looks great now!

I suggest we don't merge it before we're able to add integration tests. I already started working on it so I will push them directly to this branch and ping you when I'm done. 

Can't wait to land this feature!

kevingimbel commented:
@lucapette Exciting! I can't wait either, it's been a lot of fun working on this. :)

lucapette commented:
@kevingimbel I finally found some time to work on the feature. I tried to simplify the API as much as I could (therefore did some code removing) and introduced the integration tests we talked about. I'm satisfied with the coverage and the current status of the feature so I'm going to merge it as it is now 🎉 

It's going to be hard to make this work with #48 but I can't wait to give that a try!
